### PR TITLE
[optimizer] Merge multiple batched-add nodes into a single large node. 

### PR DESCRIPTION
**Pull request by nadavrot**

### CODE REVIEW

File: tests/unittests/OperatorTest.cpp [hunk 1]
```diff
@@ -1648,6 +1649,47 @@ TEST_P(InterpOnly, IntLookupTable) {
   }
 }
 
+/// Check that the sequence of extract-batchedadd-concat works.
+TEST_P(Operator, testBatchAdd) {
+  unsigned numSlices = 10;
+  auto *input =
```
Comment (rdzhabarov) [inline tests/unittests/OperatorTest.cpp]: can you add a test for quantized batchAdd for completeness

File: lib/Optimizer/GraphOptimizer.cpp [hunk 2]
```diff
@@ -544,6 +544,183 @@ static void mergeMatMul(Function *F) {
...
+      continue;
+    }
+
+    // We found a sequence of batched-add-slice that cover the input tensor. We
+    // can transform the graph and create one big batched-add.
+    std::vector<Node *> newSlices;
+    SliceNode *S = llvm::cast<SliceNode>(order[0]);
+    auto *BA = F->createBatchedAdd("mergedBA", S->getInput(), it.first);
+
+    // Create the new slices.
```
Comment (jfix71) [inline lib/Optimizer/GraphOptimizer.cpp]: nit: comment could be better, i.e. say that these are slices of the newly created merged batched add which will replace the original individual batched adds.

File: lib/Optimizer/GraphOptimizer.cpp [hunk 3]
```diff
@@ -544,6 +544,183 @@ static void mergeMatMul(Function *F) {
...
+    for (int i = 0, e = order.size(); i < e; i++) {
+      auto *orig = order[i];
+      newSlices.push_back(F->createSlice(orig->getName(), BA, orig->getStart(),
+                                         orig->getType()));
+    }
+
+    // Update all of our batched add nodes to use a slice from the vectorized
+    // batched add.
```
Comment (jfix71) [inline lib/Optimizer/GraphOptimizer.cpp]: nit: This comment is a bit confusing -- If I'm understanding correctly, might be better to say something like: "Replace the original individual batched adds with corresponding slices from the new merged batch add."

File: tests/unittests/OperatorTest.cpp (hunk 1 above)
Comment (nadavrot) [reply tests/unittests/OperatorTest.cpp]: Done.

**Comment by nadavrot** [comment]
Thanks for the review. I've updated the optimization (with more comments, better names, etc).

**Comment by rdzhabarov** [comment]
Nice!

**Pull request merged**